ユーザー管理機能導入

hiyak423 opened the pull request:
# what
ユーザー管理機能を実装

# why
ユーザーの登録、ログイン、ログアウトを可能にするため

# 以下 動作確認用の動画、画像
必要な情報を適切に入力して「会員登録」ボタンを押すと、ユーザーの新規登録ができる動画
https://gyazo.com/2385d28569c0fcf93a50d095b604cf0c

入力に問題のある状態で「会員登録」ボタンが押された際、ユーザーの新規登録はできず、エラーメッセージが出力される動画
https://gyazo.com/91ded327589c887048d46afbeb5cc198

必要な情報を適切に入力して「ログイン」ボタンを押すと、ユーザーのログインができる動画
https://gyazo.com/784de8418aa45f0e8231b29ac0827fc5

入力に問題のある状態で「ログイン」ボタンが押された際、ユーザーのログインはできず、エラーメッセージが出力される動画
https://gyazo.com/9e8dbc6c97d3015a5862a40396f67acb

ログアウトに成功する動画
https://gyazo.com/efe886bdece86e266590e3dc07b7be6c

テストに成功する画像
https://gyazo.com/371810010e44bb48880146b691263429

以上

hiyak423 committed:
devise gem 追加

hiyak423 committed:
usersテーブル設計

hiyak423 committed:
ビューファイル追加

hiyak423 committed:
新規登録ページ 実装

hiyak423 committed:
CSSファイル追加

hiyak423 committed:
ログインページ表示

hiyak423 committed:
ユーザー登録、ログイン機能実装

hiyak423 committed:
viewテンプレート適用

hiyak423 committed:
バリデーションの設定

hiyak423 committed:
エラーメッセージを表示

hiyak423 committed:
バリデーション設定変更

passwordは半角英数字混合、6字以上に設定するため

hiyak423 committed:
gem Rspec導入

hiyak423 committed:
gem追加 Rspec, Faker

テストコード実施のため

hiyak423 committed:
バリデーション修正

deviseにデフォルトで設定済みのため

hiyak423 committed:
factorybotを設定

hiyak423 committed:
単体テストコード実施

バリデーションの機能検証のため

hiyak423 committed:
rubocop実行

コードの整理　課題提出のため

hiyak423 committed:
エラーメッセージ修正

公式と同じようにするため

hiyak423 committed:
テストコード修正

エラーメッセージを追加したため

hiyak423 committed:
rubocop 実行

TakashiKozuma commented on app/models/user.rb:
```diff
@@ -0,0 +1,25 @@
+         :recoverable, :rememberable, :validatable
+
+  has_many :items
+  has_many :orders
```
```rb
  has_many :items
  has_many :orders
```
これらは商品出品や購入機能の実装なので、一旦コメントアウトし、それぞれの機能の実装時にコメントアウトを外すようにしましょう。

TakashiKozuma commented on app/models/user.rb:
```diff
@@ -0,0 +1,25 @@
+    PASSWORD_REGEX = /\A(?=.*?[a-z])(?=.*?\d)[a-z\d]+\z/i
+    validates_format_of :password, with: PASSWORD_REGEX, message: 'Include both letters and numbers'
+    validates :last_name, format: { with: /\A[ぁ-んァ-ン一-龥]/, message: 'Input full-width characters' }
+    validates :first_name, format: { with: /\A[ぁ-んァ-ン一-龥]/, message: 'Input full-width characters' }
```
```rb
    validates :last_name, format: { with: /\A[ぁ-んァ-ン一-龥]/, message: 'Input full-width characters' }
    validates :first_name, format: { with: /\A[ぁ-んァ-ン一-龥]/, message: 'Input full-width characters' }
```
現状の正規表現だと「佐々木」の「々」や「市ヶ谷」の「ヶ」のような一部の文字が登録できない状態かと存じますので、それらも登録できるように修正しましょう。

TakashiKozuma commented on db/migrate/20230802032102_devise_create_users.rb:
```diff
@@ -0,0 +1,49 @@
+      t.string :first_name_kana,    null: false
+      t.string :last_name_kana,     null: false
+      t.date   :birthday,           null: false
+      t.string :email,              null: false, default: "", unique: true
```
```rb
t.string :email, null: false, default: "", unique: true
```
`unique: true`はdeviseの機能でデフォルトで実装ができているので、マイグレーションファイルからは削除しておきましょう。

TakashiKozuma commented on spec/factories/users.rb:
```diff
@@ -0,0 +1,13 @@
+    last_name_kana            { 'ヤマ' }
+    birthday                  { Faker::Date.birthday }
+    email                     { Faker::Internet.unique.email }
+    password                  { Faker::Internet.password(min_length: 6) }
```
```rb
password { Faker::Internet.password(min_length: 6) }
```
現状だとpasswordが英字のみや、数字のみで出力される可能性があるので、必ず英数字混合になるように修正しましょう。

TakashiKozuma commented on spec/models/user_spec.rb:
```diff
@@ -0,0 +1,81 @@
+        @user.valid?
+        expect(@user.errors.full_messages).to include("Password confirmation doesn't match Password")
+      end
+    end
```
下記の実装条件を確認するテストも追加しましょう。

- メールアドレスが一意性であること
- メールアドレスは、@を含む必要があること
- パスワードは、6文字以上での入力が必須であること
- パスワードは、半角英数字混合での入力が必須であること
- お名前(全角)は、名字と名前がそれぞれ必須であること
- お名前(全角)は、名字と名前がそれぞれ必須であること

TakashiKozuma reviewed (commented):
レビュー依頼ありがとうございます！
コメントを致しましたので、ご確認お願いいたします。

修正後はコミット・プッシュのし忘れがないかご確認の上、再度フォームからレビュー依頼をお願い致します。

hiyak423 committed:
コメントアウト

未実装のため

hiyak423 committed:
バリデーション修正 　氏名（全角）

「々」や「ヶ] を登録可能にするため

hiyak423 committed:
マイグレーションファイル修正

emailのユニーク設定は、デフォルトで実装済みのため削除

hiyak423 committed:
Factorybot 生成値修正

passwordを英数字混合にするため

hiyak423 committed:
テスト項目の追加

以下が不足していたため
・メールアドレスが一意性であること
・メールアドレスは、@を含む必要があること
・パスワードは、6文字以上での入力が必須であること
・パスワードは、半角英数字混合での入力が必須であること
・お名前(全角)は、名字と名前がそれぞれ必須であること
・お名前(全角)は、名字と名前がそれぞれ必須であること

hiyak423 commented:
TakashiKozuma さん
ご確認いただきありがとうございました。
ご指摘いただいた箇所を修正しました。ご確認をお願いいたします。

テストに成功する画像です
https://gyazo.com/cf06f8bab4ea3cab248cb681490eaba5

TakashiKozuma commented on app/models/user.rb:
```diff
@@ -0,0 +1,26 @@
+    validates_format_of :password, with: PASSWORD_REGEX, message: 'is invalid. Include both letters and numbers'
+
+    validates :last_name, format: { with: /\A[ぁ-んァ-ン一-龠々]/, message: 'Input full-width characters' }
+    validates :first_name, format: { with: /\A[ぁ-んァ-ン一-龠々]/, message: 'Input full-width characters' }
```
```rb 
    validates :last_name, format: { with: /\A[ぁ-んァ-ン一-龠々]/, message: 'Input full-width characters' }
    validates :first_name, format: { with: /\A[ぁ-んァ-ン一-龠々]/, message: 'Input full-width characters' }
```
現状の正規表現だと「佐々木」の「々」のような一部の文字が登録できない状態かと存じますので、それらも登録できるように修正しましょう。

TakashiKozuma commented on spec/models/user_spec.rb:
```diff
@@ -0,0 +1,118 @@
+        @user.password_confirmation = '123456'
+        @user.valid?
+        expect(@user.errors.full_messages).to include('Password is invalid. Include both letters and numbers')
+      end
```
```rb
      it 'パスワードが半角英数字混合ではない' do
        @user.password = '123456'
        @user.password_confirmation = '123456'
        @user.valid?
        expect(@user.errors.full_messages).to include('Password is invalid. Include both letters and numbers')
```
`半角英数字混合でないと登録できないこと`をテストするには下記の3点をテストする必要があります。
足りていないテストを追加しましょう。
`半角英字のみでは登録できないこと`
`半角数字のみでは登録できないこと`
`全角文字を含むパスワードでは登録できない`

TakashiKozuma commented on spec/models/user_spec.rb:
```diff
@@ -0,0 +1,118 @@
+        @user.valid?
+        expect(@user.errors.full_messages).to include('Email is invalid')
+      end
+    end
```
下記を確認するテストも追加しましょう。

- お名前(全角)は、名字と名前がそれぞれ必須であること
- お名前(全角)は、名字と名前がそれぞれ必須であること

TakashiKozuma reviewed (commented):
レビュー依頼ありがとうございます！
修正点確認致しました。
コメントを追記いたしました。
ご確認お願いいたします。

修正後はコミット・プッシュのし忘れがないかご確認の上、再度フォームからレビュー依頼をお願い致します。

hiyak423 committed:
テスト項目追加

不足分の追加のため
・氏名欄の記入が必須であること
・パスワードが英数字混合の入力が必須であること

hiyak423 committed:
正規表現の修正

氏名（全角）で登録できない文字があったため

hiyak423 committed:
Update table.drawio

hiyak423 commented:
ありがとうございます。ご指摘部分の修正を行いました。
ご確認のほどよろしくお願いいたします。

テストの成功画像です
https://gyazo.com/ab00a6de9a26622434f168e677e87ab4

minachanchan reviewed (commented):
レビューのご依頼ありがとうございます。

実装内容を確認させていただきました。
**LGTMとさせていただきます。実装お疲れ様でした！**

このブランチをマージし、masterブランチ（もしくはmainブランチ）でプルしましょう。
内容が反映されていることを確認し、次の実装・学習に入りましょう！
引き続き応援しております！
